Describe the Form - Validation demo (#664)
DevExpress/devextreme-demos: JSDemos/Demos/Form/Validation/description.md

> ﻿This demo shows how to validate the Form editors in simple items. To apply validation rules to an item, specify them in the [validationRules[]](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/) array. Specify [type](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/RequiredRule/#type) and other properties for each rule. 

Reviewer: I removed the mentions of simple items because the reader might be unfamiliar with the term, and we never use it further. The replacement of the first "specify" to "declare" eliminates repetition.

```suggestion
﻿This demo shows how to validate Form editors. To apply validation rules to an editor, declare them in the [validationRules[]](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/) array. Specify [type](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/RequiredRule/#type) and other properties for each rule. 
```

> To submit a Form, add a [Button Item](/Documentation/ApiReference/UI_Components/dxForm/Item_Types/ButtonItem/) and set its [useSubmitBehavior](/Documentation/ApiReference/UI_Components/dxButton/Configuration/#useSubmitBehavior) property to **true**. When users click the button, the Form validates all editors that belong to the same group. In this demo, all editors are wrapped into the `customerData` [validationGroup](/Documentation/ApiReference/UI_Components/dxForm/Configuration/#validationGroup). The Form can be submitted to a server only if input validation is successful. Set the [showValidationSummary](/Documentation/ApiReference/UI_Components/dxForm/Configuration/#showValidationSummary) property to **true** to display all validation errors at the bottom of the Form. 

Reviewer: Taking the next paragraph into account, to submit the form, you need to perform three steps:

1. Wrap our Form in the HTML form.
2. Add a button item.
3. Set useSubmitBehavior.

I believe a list is a good choice here. But then we end up with an ambiguity: when we say "submit the form", do we mean "submit the HTML form" or "submit our Form component". To resolve this ambiguity, I replaced this phrase with "submit form data".

> ... add a [Button Item](/Documentation/ApiReference/UI_Components/dxForm/Item_Types/ButtonItem/)... When users click the button

This transition from button item to just button was not really clear, so I made it straight in step 2.

> ... all editors that belong to the same group. In this demo, all editors are wrapped into the `customerData` [validationGroup](/Documentation/ApiReference/UI_Components/dxForm/Configuration/#validationGroup).

I removed the repetition of "all editors" and moved the link to validationGroup to the end of the first sentence since this is where we mention the group for the first time. Also, if you look at the code, the editors aren't actually wrapped in this group - the group is just specified for them using the validationGroup property.

> Set the [showValidationSummary](/Documentation/ApiReference/UI_Components/dxForm/Configuration/#showValidationSummary) property to **true** to display ...

Repetition of "to". I used "enable" to get rid of it. 

```suggestion
To submit form data, do the following:

1. Wrap the Form component in the HTML `<form>` element.

1. Use the [Button Form Item](/Documentation/ApiReference/UI_Components/dxForm/Item_Types/ButtonItem/) to add a button to the form. This button will submit the form data.

1. Enable the button's [useSubmitBehavior](/Documentation/ApiReference/UI_Components/dxButton/Configuration/#useSubmitBehavior) property.

When users click the button, the Form validates editors that belong to the same [validationGroup](/Documentation/ApiReference/UI_Components/dxForm/Configuration/#validationGroup). In this demo, all editors belong to the `customerData` group. Form data can be submitted to a server only if input validation is successful. Enable the [showValidationSummary](/Documentation/ApiReference/UI_Components/dxForm/Configuration/#showValidationSummary) property to display all validation errors at the bottom of the Form.
```

> The **useSubmitBehavior** property requires that you wrap the dxForm in the HTML `<form>` element. You should also set the [preventDefault](/Documentation/ApiReference/Common/Object_Structures/dxEvent/Methods/#preventDefault) property to **true** to override the HTML form submit event. 

Reviewer: The information about wrapping our Form in the HTML form moved to the previous paragraph. Regarding preventDefault, it is only needed in this particular demo to stop the form from actually being submitted. I think we can drop this information altogether.

```suggestion
```

> The following validation rules are shown in this demo:  

Reviewer: Looks like this list should go after the first paragraph, in which we talk about validation rules.

- [RequiredRule](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/RequiredRule/)  
Demands that a validated editor has a value.

- [CompareRule](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/CompareRule/)    
Compares an editor value with a specified expression.

- [PatternRule](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/PatternRule/)    
Checks whether an editor value matches a specified pattern.

- [RangeRule](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/RangeRule/)    
Checks whether an editor value is in a specified range.

- [StringLengthRule](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/StringLengthRule/)  
Demands that an editor value length is in a specified range.

- [EmailRule](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/EmailRule/)    
Demands that an editor value matches the Email pattern.

- [AsyncRule](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/AsyncRule/)    

> Allows you to specify any conditions to be checked on the server. This check is performed last after all other rules pass validation checks. In this demo, it checks whether user input is equal to the `test@dx-email.com` string.

Reviewer: ```suggestion
Allows you to specify any conditions to be checked on the server. This check is performed last, after all other rules pass validation checks. In this demo, it checks whether user input is equal to the `test@dx-email.com` string.
```

Reviewer: "... it checks ..."

"It" refers to what?

﻿This demo shows how to validate Form editors. To apply validation rules to an editor, declare them in the [validationRules[]](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/) array. Specify [type](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/RequiredRule/#type) and other properties for each rule. 

The following validation rules are shown in this demo:  
- [RequiredRule](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/RequiredRule/)  

> Demands that a validated editor has a value.

Reviewer 2: Why do you use `Demands` here? I'm not sure how something, other than a person, can 'demand' something.

Author: Does the word 'Requires' fit better?

Reviewer 2: Yes

- [CompareRule](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/CompareRule/)    
Compares the editor's value to the specified expression.

- [PatternRule](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/PatternRule/)    
Checks whether an editor value matches a specified pattern.

- [RangeRule](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/RangeRule/)    
Checks whether an editor value is in a specified range.

- [StringLengthRule](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/StringLengthRule/)  

> Demands that an editor value length is in a specified range.

Reviewer 2: Same here

- [EmailRule](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/EmailRule/)    

> Demands that an editor value matches the Email pattern.

Reviewer 2: and here

- [AsyncRule](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/AsyncRule/)    
Allows you to add custom server-side validation logic. Rules of this type run last, only if all other rules passed. In this demo, an AsyncRule checks whether user input matches `test@dx-email.com`.

To submit form data, do the following:

1. Wrap the Form component in the HTML `<form>` element.

> 1. Use the [Button Form Item](/Documentation/ApiReference/UI_Components/dxForm/Item_Types/ButtonItem/) to add a button to the form. This button will submit the form data.

Reviewer 2: ```suggestion
1. Use the [Button Form Item](/Documentation/ApiReference/UI_Components/dxForm/Item_Types/ButtonItem/) to add a button to the form. This button submits the form data.
```

﻿This demo shows how to validate Form editors. To apply validation rules to an editor, declare them in the [validationRules[]](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/) array. Specify [type](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/RequiredRule/#type) and other properties for each rule. 

The following validation rules are shown in this demo:  
- [RequiredRule](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/RequiredRule/)  
Demands that a validated editor has a value.

- [CompareRule](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/CompareRule/)    

> Compares an editor value with a specified expression.

Reviewer 3: Part 1 of this comment applies to the previous item as well. Both part 1 and 2 apply to the rest of the items.

- `Compares an editor value`. I would use `the editor's value`, because once you assign a rule to an editor, you know which editor. If the rule has been set up, and it demands something. then the specific editor is known. 
- `a specified expression` - same logic here. I think `a specific expression` could work if you want to talk in general. But personally I'd choose `the specified expression`, because `a` adds uncertainty.

```suggestion
Compares the editor's value to the specified expression.
```

- [PatternRule](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/PatternRule/)    
Checks whether an editor value matches a specified pattern.

- [RangeRule](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/RangeRule/)    
Checks whether an editor value is in a specified range.

- [StringLengthRule](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/StringLengthRule/)  
Demands that an editor value length is in a specified range.

- [EmailRule](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/EmailRule/)    
Demands that an editor value matches the Email pattern.

- [AsyncRule](/Documentation/ApiReference/UI_Components/dxValidator/Validation_Rules/AsyncRule/)    

> Allows you to specify any conditions to be checked on the server. This check is performed last, after all other rules pass validation checks. In this demo, this rule checks whether user input is equal to the `test@dx-email.com` string.

Reviewer 3: `the check is performed` - passive and the word `perform` is best to avoid. 

In the end, I changed everything a little bit - just made it a bit more straightforward I guess. 

```suggestion
Allows you to add custom server-side validation logic. Rules of this type run last, only if all other rules passed. In this demo, an AsyncRule checks whether user input matches `test@dx-email.com`.
```

To submit form data, do the following:

1. Wrap the Form component in the HTML `<form>` element.

1. Use the [Button Form Item](/Documentation/ApiReference/UI_Components/dxForm/Item_Types/ButtonItem/) to add a button to the form. This button will submit the form data.

1. Enable the button's [useSubmitBehavior](/Documentation/ApiReference/UI_Components/dxButton/Configuration/#useSubmitBehavior) property.

> When users click the button, the Form validates editors that belong to the same [validationGroup](/Documentation/ApiReference/UI_Components/dxForm/Configuration/#validationGroup). In this demo, all editors belong to the `customerData` group. Form data can be submitted to a server only if input validation is successful. Enable the [showValidationSummary](/Documentation/ApiReference/UI_Components/dxForm/Configuration/#showValidationSummary) property to display all validation errors at the bottom of the Form.

Reviewer 3: `that belong to the same validationGroup`... The same as each other, or the same as the button's validationGroup, or the same as the form's validtionGroup? I'm can't quite understand how this works.

Author: I guess that the following sentence will be more clear: 
When users click the button, the Form validates all editors that belong to the same [validationGroup](/Documentation/ApiReference/UI_Components/dxForm/Configuration/#validationGroup) as this button.

Reviewer 3: Yes, that works, thanks! Maybe you'll need to adjust the second sentence too - not only mention editors, but `In this demo, all these control belong to the customerData group.`
